File: apps/ios/Anky/Core/Identity/WriterIdentity.swift
import CryptoSwift
import Foundation
import Web3Core
import web3swift

struct AnkyIdentityDescriptor: Equatable {
    let identityVersion: String
    let accountKind: String
    let chainId: UInt64
    let address: String
    let accountId: String
    let signingScheme: String
    let curve: String
    let recovery: String
    let derivationPath: String
}

struct WriterIdentity {
    static let identityVersion = "anky.base.eoa.v1"
    static let accountKind = "eoa"
    static let productionChainId: UInt64 = 8453
    static let testChainId: UInt64 = 84532
    static let derivationPath = "m/44'/60'/0'/0/0"
    static let signingScheme = "eip712"
    static let curve = "secp256k1"
    static let recovery = "bip39-english-12-word"

    private let privateKey: Data
    let chainId: UInt64
    let address: String

    var accountId: String {
        address
    }

    var descriptor: AnkyIdentityDescriptor {
        AnkyIdentityDescriptor(
            identityVersion: Self.identityVersion,
            accountKind: Self.accountKind,
            chainId: chainId,
            address: address,
            accountId: accountId,
            signingScheme: Self.signingScheme,
            curve: Self.curve,
            recovery: Self.recovery,
            derivationPath: Self.derivationPath
        )
    }

    init(recoveryPhrase: RecoveryPhrase, chainId: UInt64 = Self.productionChainId) throws {
        guard let seed = BIP39.seedFromMmemonics(recoveryPhrase.text),
              let rootNode = HDNode(seed: seed),
              let derivedNode = rootNode.derive(path: Self.derivationPath, derivePrivateKey: true),
              let privateKey = derivedNode.privateKey else {
            throw WriterIdentityError.couldNotDeriveBaseIdentity
        }

        try self.init(privateKey: privateKey, chainId: chainId)
    }

    private init(privateKey: Data, chainId: UInt64) throws {
        guard SECP256K1.verifyPrivateKey(privateKey: privateKey),
              let secp256k1PublicKey = SECP256K1.privateToPublic(privateKey: privateKey, compressed: false),
              let ethereumAddress = Utilities.publicToAddress(secp256k1PublicKey) else {
            throw WriterIdentityError.couldNotDeriveBaseIdentity
        }

        self.privateKey = privateKey
        self.chainId = chainId
        self.address = ethereumAddress.address
    }

    static func generate() -> WriterIdentity {
        let generated = try! generateRecoveryIdentity()
        return generated.identity
    }

    static func generateRecoveryIdentity(chainId: UInt64 = Self.productionChainId) throws -> (identity: WriterIdentity, phrase: RecoveryPhrase) {
        let phrase = try RecoveryPhrase.generate()
        return (try WriterIdentity(recoveryPhrase: phrase, chainId: chainId), phrase)
    }

    static func checksumAddress(_ address: String) throws -> String {
        guard let ethereumAddress = EthereumAddress(address) else {
            throw WriterIdentityError.invalidAddress
        }
        return ethereumAddress.address
    }

    func signDigest(_ digest: Data) throws -> String {
        guard digest.count == 32 else {
            throw WriterIdentityError.invalidDigest
        }
        let signature = SECP256K1.signForRecovery(hash: digest, privateKey: privateKey).serializedSignature
        guard let signature else {
            throw WriterIdentityError.couldNotSign
        }
        return signature.toHexString().addHexPrefix()
    }

    static func recoverAddress(digest: Data, signature: String) -> String? {
        guard let signatureData = Data.fromHex(signature),
              let address = Utilities.hashECRecover(hash: digest, signature: signatureData) else {
            return nil
        }
        return address.address
    }
}

enum WriterIdentityError: Error, Equatable {
    case couldNotDeriveBaseIdentity
    case invalidAddress
    case invalidDigest
    case couldNotSign
}

File: apps/android/app/src/main/java/inc/anky/android/core/identity/WriterIdentity.kt
package inc.anky.android.core.identity

import org.web3j.crypto.Bip32ECKeyPair
import org.web3j.crypto.Credentials
import org.web3j.crypto.Keys
import org.web3j.crypto.MnemonicUtils
import org.web3j.crypto.Sign

const val AnkyBaseIdentityVersion = "anky.base.eoa.v1"
const val AnkyBaseMainnetChainId = 8453L
const val AnkyBaseSepoliaChainId = 84532L
const val AnkyBaseDerivationPath = "m/44'/60'/0'/0/0"

data class AnkyIdentityDescriptor(
    val identityVersion: String = AnkyBaseIdentityVersion,
    val accountKind: String = "eoa",
    val chainId: Long = AnkyBaseMainnetChainId,
    val accountId: String,
    val address: String,
    val signingScheme: String = "eip712",
    val curve: String = "secp256k1",
    val recovery: String = "bip39-english-12-word",
    val derivationPath: String = AnkyBaseDerivationPath,
)

class WriterIdentity private constructor(
    private val keyPair: Bip32ECKeyPair,
    val chainId: Long = AnkyBaseMainnetChainId,
) {
    private val credentials: Credentials = Credentials.create(keyPair)
    val address: String = Keys.toChecksumAddress(credentials.address)
    val accountId: String = address
    val descriptor: AnkyIdentityDescriptor = AnkyIdentityDescriptor(
        chainId = chainId,
        accountId = accountId,
        address = address,
    )

    fun signDigest(digest32: ByteArray): String {
        require(digest32.size == 32) { "EIP-712 digest must be 32 bytes." }
        val signature = Sign.signMessage(digest32, keyPair, false)
        return "0x" +
            signature.r.hex() +
            signature.s.hex() +
            signature.v.hex()
    }

    companion object {
        const val IdentityVersion = AnkyBaseIdentityVersion
        const val BaseMainnetChainId = AnkyBaseMainnetChainId
        const val BaseSepoliaChainId = AnkyBaseSepoliaChainId
        const val DerivationPath = AnkyBaseDerivationPath

        private val HardenedBit = Bip32ECKeyPair.HARDENED_BIT
        private val BaseDerivationPath = intArrayOf(
            44 or HardenedBit,
            60 or HardenedBit,
            0 or HardenedBit,
            0,
            0,
        )

        fun fromRecoveryPhrase(phrase: RecoveryPhrase, chainId: Long = BaseMainnetChainId): WriterIdentity {
            val seed = MnemonicUtils.generateSeed(phrase.text, "")
            val master = Bip32ECKeyPair.generateKeyPair(seed)
            return WriterIdentity(Bip32ECKeyPair.deriveKeyPair(master, BaseDerivationPath), chainId)
        }

        fun generateRecoveryIdentity(): Pair<WriterIdentity, RecoveryPhrase> {
            val phrase = RecoveryPhrase.generate()
            return fromRecoveryPhrase(phrase) to phrase
        }
    }
}

private fun ByteArray.hex(): String = joinToString(separator = "") { "%02x".format(it) }
